### DISCUSSION TITLE
Consider adding opt-in support for parsing XML 1.1

### DISCUSSION THREAD

**Discussion by rossj** [rgrove/parse-xml#19] [Ideas]
Hi there,

I'm across some XML files that contain some hex references of control characters, e.g. `<a>hello&#x7;</a>`. Currently, these files give an error parsing with this library due to [the check on line 400](https://github.com/rgrove/parse-xml/blob/v2.0.3/src/index.js#L400).

What are your thoughts on relaxing this check, or perhaps having an explicit opt-in option, to allow explicit character references in this control character range?

Thanks!

**Comment by rgrove** [comment]
Hmm. Based on my understanding of [Section 2.2 of XML 1.0 5th ed.](https://www.w3.org/TR/2008/REC-xml-20081126/#charsets), control characters are invalid in XML documents even when represented by a character reference. 

Just in case I misinterpreted the spec, I tried parsing `<a>hello&#x7;</a>` with libxml2. It produced a parser error:

```
parser error : xmlParseCharRef: invalid xmlChar value 7
<a>hello&#x7;</a>
             ^
```

I'd prefer not to relax this check, even via an option, since it seems like doing so would result in a parser that consumes invalid documents, and my goal with parse-xml is to conform to the spec as closely as is practical (minus the unsafe parts). 

Of course, you should feel free to maintain this change yourself in a fork if it would be useful to you!

**Comment by rossj** [comment]
Thank you for your feedback. It seems that my files do conform to XML 1.1, as [Section 2.2. of XML 1.1](https://www.w3.org/TR/2006/REC-xml11-20060816/#NT-Char) does expand the char class to allow these additional control characters. My understanding is that XML 1.1 is not very common or widely implemented, so I'll just try some changes in my fork.

Thanks again.

**Comment by rgrove** [comment]
Ah, interesting.

My reason for not implementing an XML 1.1 parser is basically that I came across [this old excerpt](http://www.cafeconleche.org/books/effectivexml/chapters/03.html) from the book _Effective XML_, which pretty strongly recommends not bothering with XML 1.1:

> Everything you need to know about XML 1.1 can be summed up in two rules:
>
> 1.  Don't use it.
>
> 2.  (For experts only) If you speak Mongolian, Yi, Cambodian, Amharic, Dhivehi, Burmese or a very few other languages and you want to write your markup (not your text but your markup) in these languages, then you can set the version attribute of the XML declaration to 1.1. Otherwise, refer to rule 1.

But I'm not completely opposed to the idea of adding XML 1.1 parsing capability as an opt-in feature via the `<?xml>` directive and/or a runtime option, if it's something that would actually be useful and if it could be done without adding significant complexity or code weight.
